README.md
[![R-CMD-check](https://github.com/daeyounglim/adaptIVPT/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/daeyounglim/adaptIVPT/actions/workflows/R-CMD-check.yaml)


## adaptIVPT

+ `adaptIVPT` is an R package that will help you design and analyze in vitro permeation test (IVPT) data (currently supports parallel replicate design and balanced data).
+ This package contains 4 functions:
    * `msabe` for hypothesis testing (mixed-scaled average bioequivalence)
    * `prms` for calculating passing rates (passing rate mixed scaled)
    * `rss` for reestimating sample size (hence adaptive design)
    * `PRsurface` for plotting passing-rate surfaces (passing-rate surface)

### References
+ [Draft Guidance on Generic 5% Acyclovir Cream](https://www.accessdata.fda.gov/drugsatfda_docs/psg/Acyclovir_topical%20cream_RLD%2021478_RV12-16.pdf)
+ [In Vitro Permeation Test Studies for Topical Drug Products Submitted in ANDAs](https://www.fdanews.com/ext/resources/files/2022/10-21-22-InVitroPermeationTestStudiesdraftguidance.pdf?1666376063)
+ Davit, B. M., Chen, M. L., Conner, D. P., Haidar, S. H., Kim, S., Lee, C. H., Lionberger, R. A., Makhlouf, F. T., Nwakama, P. E., Patel, D. T., Schuirmann, D. J., & Yu, L. X. (2012). Implementation of a reference-scaled average bioequivalence approach for highly variable generic drug products by the US Food and Drug Administration. The AAPS journal, 14(4), 915-924. ([link](https://doi.org/10.1208/s12248-012-9406-x))
+ Potvin, D., DiLiberti, C. E., Hauck, W. W., Parr, A. F., Schuirmann, D. J., & Smith, R. A. (2008). Sequential design approaches for bioequivalence studies with crossover designs. Pharmaceutical Statistics: The Journal of Applied Statistics in the Pharmaceutical Industry, 7(4), 245-262. ([link](https://doi.org/10.1002/pst.294))

## Installation
```r
install.packages("adaptIVPT")
```

## Getting help
If you encounter a clear bug, please file an issue with a minimal reproducible example on [GitHub](https://github.com/daeyounglim/adaptIVPT/issues). For questions and other discussion, please email the [maintainer](daeyoung.lim@uconn.edu).


## Authors
+ Daeyoung Lim <daeyoung.lim@uconn.edu>
+ Elena Rantou <Elena.Rantou@fda.hhs.gov>
+ Jessica Kim <Jessica.Kim@fda.hhs.gov>
+ Sungwoo Choi <Sungwoo.Choi@fda.hhs.gov>
+ Nam Hee Choi <NamHee.Choi@fda.hhs.gov>
+ Stella Grosser <Stella.Grosser@fda.hhs.gov>

## ⚠️ Disclaimer
This package reflects the views of the authors and ***should not*** be construed to represent the FDA's views or policies.

NEWS.md
# adaptIVPT v 1.0.0
+ `msabe`
+ `rss`
+ `prms`
+ `PRsurface`

Currently, only supporting balanced data and parallel replicate design.

## New features
+ The first version of **adaptIVPT** providing functions that help design and analyze IVPT studies.

R/prms.R
#' Compute the passing rate for the mixed scaling approach in bioequivalence (BE) studies
#' 
#' This function runs Monte Carlo simulations to compute the passing rate (PR) of the mixed scaling (MS) approach.
#' @author Daeyoung Lim, \email{daeyoung.lim@uconn.edu}
#' @param n The number of donors in each simulation.
#' @param r The number of replicates from each donor for each simulated dataset.
#' @param params (Optional) The list of true parameters to be assumed in data generation.
#' 
#' + `sigma_W0` - A regulatory constant set by the FDA. Defaults to 0.25.
#' + `sigma_WT` - The true standard deviation of the test formulation population.
#' + `sigma_WR` - The true standard deviation of the reference formulation population.
#' + `GMR` - The geometric mean ratio of the test and reference values of the pharmacokinetic measures (e.g., Jmax or AUC). If the test-formulation measure is greater than that of the reference formulation, then GMR is typically set to 1.05, which is the initial value of this function. If the reference-formulation measure is bigger, then GMR is typically 0.95. Defaults to 0.95.
#' + `m` - Another regulatory constant that determines the bounds within which the estimated GMR should fall for bioequivalence to be established. Defaults to 1.25, representing 80-125% average BE limits, which is the FDA recommendation.
#' + `sig_level` - The significance level (alpha-level). Defaults to 0.05.
#' 
#' @param nsim (Optional) The number of total simulations to be conducted. Defaults to 1,000.
#' @param ncores (Optional) The number of CPU cores to use for parallel processing (OpenMP). If R hasn't been installed with OpenMP configured, this will not take effect. When OpenMP is available, it should not exceed the number of existing cores. If unspecified, it will default to 2 cores or the number of existing cores, whichever is smaller.
#' @references
#' Davit, B. M., Chen, M. L., Conner, D. P., Haidar, S. H., Kim, S., Lee, C. H., Lionberger, R. A., Makhlouf, F. T., Nwakama, P. E., Patel, D. T., Schuirmann, D. J., & Yu, L. X. (2012). Implementation of a reference-scaled average bioequivalence approach for highly variable generic drug products by the US Food and Drug Administration. The AAPS journal, 14(4), 915-924.
#' 
#' @return A list of lists
#' 
#' + `parameters` - A list of true parameter settings.
#' + `passing_rate` - The estimated passing rate.
#' + `runtime` - The total elapsed time charged for the execution of the program.
#' 
#' @examples
#' out <- prms(10, 6, nsim = 2)
#' 
#' @md
#' @export
prms <- function(n, r, params = list(), nsim = 1000, ncores = NULL) {
    parameters <- list(sigma_W0 = 0.25, sigma_WT = 0.5, sigma_WR = 0.5, GMR = 0.95,
                       m = 1.25, sig_level = 0.05)
    parameters[names(params)] <- params
    sigma_W0 <- parameters$sigma_W0
    sigma_WR <- parameters$sigma_WR
    sigma_WT <- parameters$sigma_WT
    GMR <- parameters$GMR
    m <- parameters$m
    sig_level <- parameters$sig_level

    if (!is.null(ncores)) {
        ncores_ <- parallel::detectCores()
        if (ncores > ncores_) {
            stop(paste0("The number of cores cannot exceed ", ncores_))
        }
    } else {
        ncores <- min(2, parallel::detectCores())
    }

    runtime <- system.time({
        fout <- .Call(
            `_adaptIVPT_prms`,
            as.integer(n),
            as.integer(r),
            as.double(sigma_WT^2),
            as.double(sigma_WR^2),
            as.double(log(GMR)),
            as.double(m),
            as.double(sigma_W0),
            as.double(sig_level),
            as.integer(nsim),
            as.integer(ncores))
    })[3]
    list(parameters = parameters,
         passing_rate = fout,
         runtime = runtime)
}

src/misc.cpp
#include <iostream>
#include <cmath>
#include "misc.h"


double sim1_mixedscaling(const int &n,
                          const int &r,
                          const double &m,
                          const double &theta,
                          const double &t_cutoff, // = t_{(1-\alpha, n-1)}
                          const double &chi2_cutoff,
                          const double &S2_I,
                          const double &S2_WR,
                          const double Ibar) { // = chi2_{(1-\alpha, n*(r-1))}
    double out = 0.0;
    if (std::sqrt(S2_WR) >= 0.294) {
        double X = Ibar * Ibar - S2_I / static_cast<double>(n);
        double Y = -theta * S2_WR;
        double X_beta = std::pow(std::abs(Ibar) + t_cutoff * std::sqrt(S2_I / static_cast<double>(n)), 2.0);
        double Y_beta = -theta * static_cast<double>((r-1)*n) * S2_WR / chi2_cutoff;
        double X_diff = X_beta - X;
        double Y_diff = Y_beta - Y;
        double V = X_diff * std::abs(X_diff) + Y_diff * std::abs(Y_diff);
        double SABE_upperbound = X + Y + sgn(V) * std::sqrt(std::abs(V));
        if (SABE_upperbound <= 0.0 && std::abs(Ibar) < std::log(m)) {
            out = 1.0;
        }
    } else {
        double TOST_upperbound = Ibar + t_cutoff * std::sqrt(S2_I / static_cast<double>(n));
        double TOST_lowerbound = Ibar - t_cutoff * std::sqrt(S2_I / static_cast<double>(n));
        if (TOST_lowerbound >= -std::log(m) && TOST_upperbound <= std::log(m)) {
            out = 1.0;
        }
    }
    return out;
}

src/prms.cpp
#include <iostream>
#include <cmath>
#include <RcppArmadillo.h>
#include <Rmath.h>
#include <Rdefines.h>
#include <algorithm>
#include <iterator>
#include <progress.hpp>
#include <progress_bar.hpp>
#include "misc.h"
#ifdef _OPENMP
  #include <omp.h>
#endif
// [[Rcpp::plugins(openmp)]]
// [[Rcpp::plugins(cpp11)]] 
// [[Rcpp::depends(RcppArmadillo,RcppProgress)]]


/********************************
n is # of donors
r is # of skin replicates
Test:      n by r matrix
Reference: n by r matrix
********************************/
// [[Rcpp::export]]
arma::vec msabe(const arma::mat &Test,
           const arma::mat &Reference,
           const double &m,
           const double &sigma_W0,
           const double &sig_level) {
    int n = Test.n_rows;
    int r = Test.n_cols;
    const double theta = std::pow(std::log(m) / sigma_W0, 2.0);
    const double t_cutoff = R::qt(1.0 - sig_level, static_cast<double>(n-1), 1, 0);
    const double chi2_cutoff = R::qnchisq(1.0 - sig_level, static_cast<double>(n*(r-1)), 0.0, 1, 0);
    arma::vec Rj_bar = arma::mean(Reference, 1);
    double S2_WR = arma::accu(arma::square(Reference.each_col() - Rj_bar)) / static_cast<double>(n*(r-1));
    arma::vec Ij = arma::mean(Test - Reference, 1);
    double Ibar = arma::mean(Ij);
    double S2_I = arma::var(Ij);

    double out = 0.0;
    double upperbound = 0.0;
    double lowerbound = 0.0;
    if (std::sqrt(S2_WR) >= 0.294) {
        double X = Ibar * Ibar - S2_I / static_cast<double>(n);
        double Y = -theta * S2_WR;
        double X_beta = std::pow(std::abs(Ibar) + t_cutoff * std::sqrt(S2_I / static_cast<double>(n)), 2.0);
        double Y_beta = -theta * static_cast<double>((r-1)*n) * S2_WR / chi2_cutoff;
        double X_diff = X_beta - X;
        double Y_diff = Y_beta - Y;
        double V = X_diff * std::abs(X_diff) + Y_diff * std::abs(Y_diff);
        double SABE_upperbound = X + Y + sgn(V) * std::sqrt(std::abs(V));
        upperbound = SABE_upperbound;
        if (SABE_upperbound <= 0.0 && std::abs(Ibar) < std::log(m)) {
            out = 1.0;
        }
    } else {
        double TOST_upperbound = Ibar + t_cutoff * std::sqrt(S2_I / static_cast<double>(n));
        double TOST_lowerbound = Ibar - t_cutoff * std::sqrt(S2_I / static_cast<double>(n));
        upperbound = TOST_upperbound;
        lowerbound = TOST_lowerbound;
        if (TOST_lowerbound >= -std::log(m) && TOST_upperbound <= std::log(m)) {
            out = 1.0;
        }
    }
    return {out, Ibar, S2_WR, S2_I, lowerbound, upperbound};
}

// [[Rcpp::export]]
double prms(const int &n,
            const int &r,
            const double &sigma2_WT,
            const double &sigma2_WR,
            const double &logGMR,
            const double &m,
            const double &sigma_W0,
            const double &sig_level, // significance level (alpha)
            const int &nsim,
            const int &ncores) {
    const double theta = std::pow(std::log(m) / sigma_W0, 2.0);
    const double t_cutoff = R::qt(1.0 - sig_level, static_cast<double>(n-1), 1, 0);
    const double chi2_cutoff = R::qnchisq(1.0 - sig_level, static_cast<double>(n*(r-1)), 0.0, 1, 0);
    arma::vec out(nsim, arma::fill::zeros);
    arma::vec S2_WRs(nsim, arma::fill::zeros);
    arma::vec S2_Is(nsim, arma::fill::zeros);
    for (int isim = 0; isim < nsim; ++isim) {
        S2_WRs(isim) = ::Rf_rnchisq(static_cast<double>(n*(r-1)), 0.0) * sigma2_WR / static_cast<double>(n*(r-1));
        S2_Is(isim) = ::Rf_rnchisq(static_cast<double>(n-1), 0.0) * (sigma2_WR + sigma2_WT) / static_cast<double>(r*(n-1));
    }
    arma::vec Ibars(nsim, arma::fill::randn);
    Ibars *= std::sqrt((sigma2_WT + sigma2_WR) / static_cast<double>(n*r));
    Ibars += logGMR;


    #ifdef _OPENMP
    #pragma omp parallel for schedule(static) num_threads(ncores)
    #endif
    for (int isim = 0; isim < nsim; ++isim) {
        double S2_I = S2_Is(isim);
        double S2_WR = S2_WRs(isim);
        double Ibar = Ibars(isim);
        double test_function = sim1_mixedscaling(n, r, m, theta, t_cutoff, chi2_cutoff, S2_I, S2_WR, Ibar);
        out(isim) = test_function;
    }
    return arma::mean(out);
}

// [[Rcpp::export]]
int reestimate_samplesize(const int &n_init,
           const int &r,
           const double &lGMR,
           const double &S2_WR,
           const double &alpha_level,
           const double &m,
           const double &sigma_W0,
           const double &target_power,
           const int &nmax,
           const int &nsim,
           const int &ncores) {
    int n = n_init + 1;
    // Double n until the passing rate is >= target_power
    double out_old = 0.0;
    while (n < nmax) {
        out_old = prms(n, r, S2_WR, S2_WR, lGMR, m, sigma_W0, alpha_level, nsim, ncores);
        if (out_old >= target_power) {
            break;
        } else {
            n *= 2;
        }
    }
    if (n >= nmax) {
        return nmax;
    }

    // Binary search
    int n_lower = n / 2;
    int n_upper = n;
    double out = out_old;
    while (n_upper - n_lower > 1) {
        n = (n_upper + n_lower) / 2;
        out = prms(n, r, S2_WR, S2_WR, lGMR, m, sigma_W0, alpha_level, nsim, ncores);
        if (out < target_power) {
            n_lower = n;
        } else {
            n_upper = n;
        }
    }
    return n_upper;
}

R/power_plot.R
#' Plot the passing-rate curve and the passing-rate surface in IVPT
#' 
#' This function plots the power (passing-rate) curve and power (passing-rate) surface of the mixed scaling (MS) approach. A power curve shows the statistical power across different effect sizes. In IVPT studies, the effect size is captured by the difference between the means of log-measurements of the test and reference products (i.e., logGMR). For the passing-rate surface, the corresponding function considers different values of the standard deviation.
#' @author Daeyoung Lim, \email{daeyoung.lim@uconn.edu}
#' @param n The number of donors in each simulation.
#' @param r The number of replicates from each donor for each simulated dataset.
#' @param observed_GMR The observed (estimated) GMR of the user's data. Along with the observed sigmaWR, the corresponding passing rate will be displayed in the 3D plot as a vertical line parallel to the z-axis.
#' @param observed_sigmaWR The observed (estimated) sigmaWR of the user's data. Along with the observed GMR, the corresponding passing rate will be displayed in the 3D plot as a vertical line parallel to the z-axis.
#' @param GMR_grid The grid of GMR values to be used for plotting the 3D surface of passing rates.
#' @param sigmaWR_grid The grid of sigmaWR values to be used for plotting the 3D surface of passing rates.
#' @param params (Optional) The list of true parameters to be assumed in data generation.
#' 
#' + `sigma_W0` - A regulatory constant set by the FDA. Defaults to 0.25.
#' + `sigma_WT` - The true standard deviation of the test formulation population.
#' + `sigma_WR` - The true standard deviation of the reference formulation population.
#' + `GMR` - The geometric mean ratio of the test and reference values of the pharmacokinetic measures (e.g., Jmax or AUC). If the test-formulation measure is greater than that of the reference formulation, then GMR is typically set to 1.05, which is the initial value of this function. If the reference-formulation measure is bigger, then GMR is typically 0.95. Defaults to 0.95.
#' + `m` - Another regulatory constant that determines the bounds within which the estimated GMR should fall for bioequivalence to be established. Defaults to 1.25, representing 80-125% average BE limits, which is the FDA recommendation.
#' + `sig_level` - The significance level (alpha-level). Defaults to 0.05.
#' 
#' @param nsim (Optional) The number of total simulations to be conducted. Defaults to 1,000.
#' @param ncores (Optional) The number of CPU cores to use for parallel processing (OpenMP). If R hasn't been installed with OpenMP configured, this will not take effect. When OpenMP is available, it should not exceed the number of existing cores. If unspecified, it will default to 2 cores or the number of existing cores, whichever is smaller.
#' @param verbose (Optional) A logical value (`TRUE`/`FALSE`) indicating whether to display the progress bar.
#' @param plot (Optional) A logical value (`TRUE`/`FALSE`) indicating whether to generate a 3D interactive plot of the surface. If `FALSE`, the function will return the (x, y, z) values as a `list`.
#' @references
#' Davit, B. M., Chen, M. L., Conner, D. P., Haidar, S. H., Kim, S., Lee, C. H., Lionberger, R. A., Makhlouf, F. T., Nwakama, P. E., Patel, D. T., Schuirmann, D. J., & Yu, L. X. (2012). Implementation of a reference-scaled average bioequivalence approach for highly variable generic drug products by the US Food and Drug Administration. The AAPS journal, 14(4), 915-924.
#' 
#' @return A list
#' 
#' + `GMR` - A list of true parameter settings.
#' + `passing_rate` - The estimated passing rate.
#' + `runtime` - The total elapsed time charged for the execution of the program.
#' 
#' @examples
#' out <- PRsurface(6, 3, GMR_grid = c(0.90, 1), sigmaWR_grid = c(0.2, 0.5), nsim = 2, plot = FALSE)
#' 
#' @md
#' @importFrom rgl persp3d plot3d
#' @importFrom utils setTxtProgressBar txtProgressBar
#' @export
PRsurface <- function(n, r, observed_GMR = 0.95, observed_sigmaWR = 0.294, GMR_grid = seq(0.75, 1.30, length.out = 100), sigmaWR_grid = seq(0.2, 1, length.out = 100),
                      params = list(), nsim = 1000, ncores = NULL, verbose = FALSE, plot = TRUE) {
    parameters <- list(sigma_W0 = 0.25, sigma_WT = 0.5, sigma_WR = 0.5, GMR = 0.95,
                       m = 1.25, sig_level = 0.05)
    parameters[names(params)] <- params
    z <- matrix(0, nrow = length(GMR_grid), ncol = length(sigmaWR_grid))
    if (verbose) {
        pb <- txtProgressBar(style=3)
    }
    iprog <- 0
    nprog <- length(GMR_grid) * length(sigmaWR_grid)
    for (i in 1:length(GMR_grid)) {
        GMR <- GMR_grid[i]
        parameters$GMR <- GMR
        for (j in 1:length(sigmaWR_grid)) {
            sigmaWR <- sigmaWR_grid[j]
            parameters$sigma_WR <- parameters$sigma_WT <- sigmaWR
            z[i,j] <- prms(n, r, parameters, nsim, ncores)$passing_rate
            if (verbose) {
                iprog <- iprog + 1
                setTxtProgressBar(pb, iprog / nprog)
            }
        }
    }
    if (verbose) {
        close(pb)
    }
    if (plot) {
        persp3d(GMR_grid, sigmaWR_grid, z, main="Passing Rate Surface",
                zlab = "Passing Rate", xlab="GMR", ylab=expression(sigma[WR]),
                theta = 30, phi = 15, col = "orange", shade = 0.4)
        ml <- max(length(GMR_grid), length(sigmaWR_grid))
        observed_pr <- prms(n, r, list(sigma_WR=observed_sigmaWR, sigma_WT = observed_sigmaWR, GMR = observed_GMR))$passing_rate
        plot3d(rep(observed_GMR, ml), rep(observed_sigmaWR, ml), seq(0, observed_pr, length.out = ml), type = 'l', col="red", lwd = 2, add = TRUE)
    }
    invisible(list(GMR = GMR_grid, sigmaWR = sigmaWR_grid, z = z))
}

R/reestimate_samplesize.R
#' Reestimate the sample size for the adaptive design in bioequivalence (BE) studies using mixed criterion.
#' 
#' This function reestimates the sample size using mixed criterion required for target power, using binary search. The power (passing rate) function of mixed criterion testing lacks a closed-form expression. Thus, sample size (re-)estimation requires a binary search, after identifying an `n` where the passing rate exceeds the desired level.
#' @author Daeyoung Lim, \email{daeyoung.lim@uconn.edu}
#' @param n The number of donors in each simulation.
#' @param r The number of replicates from each donor for each simulated dataset.
#' @param S_WR The estimated standard deviation of the reference measurements. The reference-scaled average bioequivalence approach is used if S_WR > 0.249 and the average bioequivalence approach otherwise.
#' @param params (Optional) The list of true parameters to be assumed in data generation.
#' 
#' + `sigma_W0` - A regulatory constant set by the FDA. Defaults to 0.25.
#' + `GMR` - The geometric mean ratio of the test and reference values of the pharmacokinetic measures (e.g., Jmax or AUC). If the test-formulation measure is greater than that of the reference formulation, then GMR is typically set to 1.05, which is the initial value of this function. If the reference-formulation measure is bigger, then GMR is typically 0.95. Defaults to 0.95.
#' + `m` - Another regulatory constant that determines the bounds within which the estimated GMR should fall for bioequivalence to be established. Defaults to 1.25, representing 80-125% average BE limits, which is the FDA recommendation.
#' + `sig_level` - The significance level (alpha-level).
#' + `nmax` - The upper limit for sample size reestimation. If the sample size exceeds `nmax` inside estimation procedure, the function will return `nmax`.
#' + `target_power` - The threshold for power (or passing rate) for a hypothesis test to be considered powerful. Typically set at 80% and defaults to 0.8.
#' 
#' @param nsim (Optional) The number of total simulations to be conducted. Defaults to 1,000.
#' @param ncores (Optional) The number of CPU cores to use for parallel processing (OpenMP). If R hasn't been installed with OpenMP configured, this will not take effect. When OpenMP is available, it should not exceed the number of existing cores. If unspecified, it will default to 2 cores or the number of existing cores, whichever is smaller.
#' 
#' @return A list of lists
#' 
#' + `parameters` - A list of true parameter settings.
#' + `rss` - The reestimated sample size.
#' + `runtime` - The total elapsed time charged for the execution of the program.
#' 
#' @examples
#' out <- rss(10, 6, S_WR = 0.22, nsim = 2)
#' 
#' @references
#' Potvin, D., DiLiberti, C. E., Hauck, W. W., Parr, A. F., Schuirmann, D. J., & Smith, R. A. (2008). Sequential design approaches for bioequivalence studies with crossover designs. Pharmaceutical Statistics: The Journal of Applied Statistics in the Pharmaceutical Industry, 7(4), 245-262.
#' 
#' @md
#' @export
rss <- function(n, r, S_WR, params = list(), nsim = 1000, ncores = NULL) {
    parameters <- list(sigma_W0 = 0.25, GMR = 0.95,
                       m = 1.25, sig_level = 0.05, nmax = 100, target_power = 0.8)
    parameters[names(params)] <- params
    sigma_W0 <- parameters$sigma_W0
    GMR <- parameters$GMR
    m <- parameters$m
    sig_level <- parameters$sig_level
    nmax <- parameters$nmax
    target_power <- parameters$target_power

    if (!is.null(ncores)) {
        ncores_ <- parallel::detectCores()
        if (ncores > ncores_) {
            stop(paste0("The number of cores cannot exceed ", ncores_))
        }
    } else {
        ncores <- min(2, parallel::detectCores())
    }

    runtime <- system.time({
        fout <- .Call(
            `_adaptIVPT_reestimate_samplesize`,
            as.integer(n),
            as.integer(r),
            as.double(log(GMR)),
            as.double(S_WR^2),
            as.double(sig_level),
            as.double(m),
            as.double(sigma_W0),
            as.double(target_power),
            as.integer(nmax),
            as.integer(nsim),
            as.integer(ncores))
    })[3]
    list(parameters = parameters,
         rss = fout,
         runtime = runtime)
}

R/msabe.R
#' Run the mixed scaling approach in bioequivalence (BE) studies
#' 
#' This function runs hypothesis testing for bioequivalence using the mixed criterion
#' @author Daeyoung Lim, \email{daeyoung.lim@uconn.edu}
#' @param Test An n-by-r matrix of test product data. `n` is the number of donors and `r` is the number of skin section replicates.
#' @param Reference An n-by-r matrix of reference product data.
#' @param params (Optional) The list of tuning parameters for running the test.
#' 
#' + `sigma_W0` - A regulatory constant set by the FDA. Defaults to 0.25.
#' + `m` - Another regulatory constant that determines the bounds within which the estimated GMR should fall for bioequivalence to be established. Defaults to 1.25, representing 80-125% average BE limits, which is the FDA recommendation.
#' + `sig_level` - The significance level (alpha-level).
#' 
#' @references
#' Davit, B. M., Chen, M. L., Conner, D. P., Haidar, S. H., Kim, S., Lee, C. H., Lionberger, R. A., Makhlouf, F. T., Nwakama, P. E., Patel, D. T., Schuirmann, D. J., & Yu, L. X. (2012). Implementation of a reference-scaled average bioequivalence approach for highly variable generic drug products by the US Food and Drug Administration. The AAPS journal, 14(4), 915-924.
#' 
#' @return A list of lists
#' 
#' + `parameters` - A list of true parameter settings.
#' + `fout` - The test result and related estimators.
#' + `runtime` - The total elapsed time charged for the execution of the program.
#' 
#' @examples
#' n <- 6
#' r <- 3
#' Test <- matrix(runif(n*r), nrow = n, ncol = r)
#' Reference <- matrix(runif(n*r), nrow = n, ncol = r)
#' out <- msabe(Test, Reference)
#' 
#' @md
#' @export
msabe <- function(Test, Reference, params = list()) {
    parameters <- list(sigma_W0 = 0.25, m = 1.25, sig_level = 0.05)
    parameters[names(params)] <- params
    sigma_W0 <- parameters$sigma_W0
    m <- parameters$m
    sig_level <- parameters$sig_level

    runtime <- system.time({
        fout <- .Call(
            `_adaptIVPT_msabe`,
            as.matrix(Test),
            as.matrix(Reference),
            as.double(m),
            as.double(sigma_W0),
            as.double(sig_level))
    })[3]
    fout <- drop(fout)
    names(fout) <- c("test", "Ibar", "S2_WR", "S2_I", "lowerbound", "upperbound")
    out <- list(parameters = parameters,
         fout = fout,
         runtime = runtime)

    class(out) <- "msabe"
    out
}

